cinn/auto_schedule/database/database.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2022 CINN Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "cinn/auto_schedule/measure/measure.h"
+#include "cinn/auto_schedule/search_space/search_state.h"
+
+namespace cinn {
+namespace auto_schedule {
+
+// Record related data about tuning process of a measure candidate
+struct TuningRecord {
+  // the unique key to identify a task
+  std::string task_key;
+  // the cost time of the candidate executed during measure
+  double execution_cost;  // unit: us
+  // the searched candidate to be saved
+  SearchState state;
+
+  // a binary compare function that denotes when the left
+  // will be sorted in the front of the right
+  struct Compare {
+    bool operator()(const TuningRecord& lhs, const TuningRecord& rhs) const;
+  };
+};
+
+// A database supports insert or lookup historial tuning result with sepecified traits.
+// It can be implemented with a concrete storage to save/load underlying data,
+// such as memory, file, database server.
+class Database {
```

Comment:
I think the `MemoryDatabase` is just a work in progress? We finally will save the records in `MemoryDatabase` in the disk. The memory database code won't exist at that time. Could we just name a database that you don't have to rename in the future?

Reply:
Take it, I remove the `MemoryDatabase ` class and move its implement functions to the base class which saves data in memory as default